src/librustc_trait_selection/traits/fulfill.rs

```diff
@@ -318,267 +319,216 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
         let infcx = self.selcx.infcx();
 
         match obligation.predicate.kind() {
-            ty::PredicateKind::Trait(ref data, _) => {
-                let trait_obligation = obligation.with(*data);
-
-                if obligation.predicate.is_global() {
-                    // no type variables present, can use evaluation for better caching.
-                    // FIXME: consider caching errors too.
-                    if infcx.predicate_must_hold_considering_regions(&obligation) {
-                        debug!(
-                            "selecting trait `{:?}` at depth {} evaluated to holds",
-                            data, obligation.recursion_depth
-                        );
-                        return ProcessResult::Changed(vec![]);
+            ty::PredicateKind::ForAll(binder) => match binder.skip_binder().kind() {
+                ty::PredicateKind::ForAll(_) => bug!("unexpected forall"),
+                // Evaluation will discard candidates using the leak check.
+                // This means we need to pass it the bound version of our
+                // predicate.
+                &ty::PredicateKind::Atom(atom) => match atom {
+                    ty::PredicateAtom::Trait(trait_ref, _constness) => {
+                        let trait_obligation = obligation.with(Binder::bind(trait_ref));
+
+                        self.process_trait_obligation(
```

Comment:
I think that *eventually* we actually want trait *selection* to take a `TraitPredicate`, not a `PolyTraitPredicate`, so that the handling of for-all is quite uniform. 
But this PR is probably drawing the line at the right place for now.